src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -167,51 +170,50 @@ public void addSolidTransaction(Hash hash) {
-                        TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hash);
-                        Set<Hash> approvers = transaction.getApprovers(tangle).getHashes();
-                        for(Hash h: approvers) {
-                            TransactionViewModel tx = TransactionViewModel.fromHash(tangle, h);
-                            if(quietQuickSetSolid(tx)) {
-                                tx.update(tangle, "solid");
-                            } else {
-                                if (transaction.isSolid()) {
-                                    addSolidTransaction(hash);
-                                }
-                            }
-                        }
-                    } catch (Exception e) {
-                        log.error("Some error", e);
-                        // TODO: Do something, maybe, or do nothing.
-                    }
-                }
-                synchronized (cascadeSync) {
-                    if (useFirst.get()) {
-                        newSolidTransactionsTwo.clear();
-                    } else {
-                        newSolidTransactionsOne.clear();
+                propagateSolidTransactions();
+            }
+        };
+    }
+
+    void propagateSolidTransactions() {
+        Set<Hash> newSolidHashes = new HashSet<>();
+        useFirst.set(!useFirst.get());
+        synchronized (cascadeSync) {
+            if (useFirst.get()) {
+                newSolidHashes.addAll(newSolidTransactionsTwo);
+            } else {
+                newSolidHashes.addAll(newSolidTransactionsOne);
+            }
+        }
+        Iterator<Hash> cascadeIterator = newSolidHashes.iterator();
+        while(cascadeIterator.hasNext() && !shuttingDown.get()) {
+            try {
+                Hash hash = cascadeIterator.next();
+                TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hash);
+                Set<Hash> approvers = transaction.getApprovers(tangle).getHashes();
+                for(Hash h: approvers) {
+                    TransactionViewModel tx = TransactionViewModel.fromHash(tangle, h);
+                    if(quietQuickSetSolid(tx)) {
+                        tx.update(tangle, "solid");
+                        addSolidTransaction(h);
                     }
                 }
-                try {
-                    Thread.sleep(500);
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
-                }
+            } catch (Exception e) {
+                log.error("Error while propagating solidity upwards", e);
             }
-        };
+        }
+        synchronized (cascadeSync) {
+            if (useFirst.get()) {
+                newSolidTransactionsTwo.clear();
+            } else {
+                newSolidTransactionsOne.clear();
+            }
+        }
+        try {
+            Thread.sleep(500);
```

Comment:
move the sleep to the runnable? this method shouldn't have a sleepy side effect

Reply:
This was part of the old code. I only wanted to fix the bug.
I wanted to change as little as possible from the old code and see the effect.

I will open a general issue on solidification though. Because indeed the way things are written should be changed.